hammer: rbd snap rollback: restore the link to parent

asheplyakov opened the pull request:
So snapshot, flatten, rollback of a cloned image does not loose any data

Fixes: #14512
Signed-off-by: Alexey Sheplyakov <asheplyakov@mirantis.com>

dillaman commented on src/librbd/internal.cc:
```diff
@@ -444,6 +444,30 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+          const SnapInfo& si(it->second);
+          // current parent does not match the one of the snapshot being reverting to
+          bool should_update_parent = ictx->parent_md.spec != si.parent.spec ||
+                  ictx->parent_md.overlap != si.parent.overlap;
```
If the overlap changes (but parent record still exists), the cls `set_parent` method will return `-EEXISTS` when trying to update the parent record.  Therefore, you will need to `remove_parent` first.  Also need to support the case where HEAD is associated with a parent but the rollback snapshot doesn't have a parent.

dillaman commented on src/test/librbd/test_snapshot_flatten_rollback_cloned.py:
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python
```
Can you move this test to `test_rbd.py`?  Probably enough to just verify parent spec after rollback.  Would also be good to add tests for various combinations of updates (no parent->parent, parent->no parent, overlap size change, etc).

asheplyakov commented:
@dillaman

> If the overlap changes (but parent record still exists), the cls set_parent method will return -EEXISTS when trying to update the parent record. Therefore, you will need to remove_parent first.

Done.

> Also need to support the case where HEAD is associated with a parent but the rollback snapshot doesn't have a parent.

Done.

> Can you move this test to test_rbd.py?

Done.

> Probably enough to just verify parent spec after rollback. 

I've added parent info checks, however think it's important to verify if the actual data matches the expected one.

> Would also be good to add tests for various combinations of updates (no parent->parent, parent->no parent, overlap size change, etc).

I've implemented no parent -> parent, parent -> no parent tests, however I don't quite understand which operations can change the overlap size. Any hints, please?

dillaman commented:
@asheplyakov The only way to change the overlap size is to shrink the image.  So if you resize the image down past the original parent size, the overlap is now smaller.

asheplyakov commented:
@dillaman 
> The only way to change the overlap size is to shrink the image. 

I've added a test which verifies that rollback handles such a modification. Does the patch look OK now?

dillaman commented on src/librbd/internal.cc:
```diff
@@ -413,6 +413,58 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+    CephContext *cct = ictx->cct;
+    int r = 0;
+    std::map<librados::snap_t, SnapInfo>::const_iterator it = ictx->snap_info.find(snap_id);
+    assert(it != ictx->snap_info.end());
```
Probably sufficient to just return ```-ENOENT``` in this case.

asheplyakov replied:
Done

dillaman commented on src/librbd/internal.cc:
```diff
@@ -413,6 +413,58 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+    std::map<librados::snap_t, SnapInfo>::const_iterator it = ictx->snap_info.find(snap_id);
+    assert(it != ictx->snap_info.end());
+    const SnapInfo& snap_info(it->second);
+    if (snap_info.parent.spec.snap_id == CEPH_NOSNAP) {
```
Minor: while technically correct to check ```snap_id == CEPH_NOSNAP```, since the rest of librbd uses ```pool_id == -1``` for the same check, it might make it clearer (here and below)

asheplyakov replied:
Done

dillaman commented on src/librbd/internal.cc:
```diff
@@ -413,6 +413,58 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+    assert(it != ictx->snap_info.end());
+    const SnapInfo& snap_info(it->second);
+    if (snap_info.parent.spec.snap_id == CEPH_NOSNAP) {
+      if (ictx->parent_md.spec.snap_id == CEPH_NOSNAP) {
```
Minor: instead of the nested ```if```'s, can we just test:

```
if (ictx->parent_md == snap_info.parent) { 
  return 0;
}

if (ictx->parent_md.pool_id != -1) {
  ... remove parent ...
}
if (snap_info.parent.pool_id != -1) {
  ... set parent ...
} 
```

asheplyakov replied:
Done

dillaman commented on src/test/pybind/test_rbd.py:
```diff
@@ -1048,3 +1048,96 @@ def test_follower_write(self):
                 eq(data, read)
+
+
+class TestRollbackTricky(object):
```
s/TestRollbackTricky/TestCloneRollback/

asheplyakov replied:
Done

dillaman commented:
lgtm

dillaman commented:
http://pulpito.ceph.com/jdillaman-2016-05-11_07:32:22-rbd-wip-jd-testing---basic-mira/

dillaman commented:
👍 -- good to merge once 0.94.7 is released.
